crates/video/src/video.rs

```diff
@@ -0,0 +1,1089 @@
+        }
+    }
+}
+
+impl virtio_video_format_range {
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe {
+            return ::core::slice::from_raw_parts(
+                (self as *const virtio_video_format_range) as *const u8,
+                ::core::mem::size_of::<virtio_video_format_range>(),
+            );
+        };
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Default)]
+pub struct virtio_video_format_frame {
+    pub width: virtio_video_format_range,
+    pub length: virtio_video_format_range,
+    pub num_rates: Le32,
+    pub padding: Le32,
+    pub frame_rates: Vec<virtio_video_format_range>,
+}
+
+impl virtio_video_format_frame {
+    pub fn append_frame_rates(&mut self, frames: &mut Vec<virtio_video_format_range>) {
+        self.frame_rates.append(frames);
+        self.num_rates = (self.frame_rates.len() as u32).into();
+    }
+}
+
+impl ToBytes for virtio_video_format_frame {
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut ret: Vec<u8> = Vec::new();
+
+        ret.extend_from_slice(self.width.as_slice());
+        ret.extend_from_slice(self.length.as_slice());
+        ret.extend_from_slice(self.num_rates.as_slice());
+        ret.extend_from_slice(self.padding.as_slice());
+        for frame in self.frame_rates.iter() {
+            ret.extend_from_slice(frame.as_slice());
+        }
+        ret
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Default)]
+pub struct virtio_video_format_desc {
+    mask: Le64,
+    pub format: Le32,
+    pub planes_layout: Le32,
+    plane_align: Le32,
+    num_frames: Le32,
+    frames: Vec<virtio_video_format_frame>,
+}
+
+impl virtio_video_format_desc {
+    pub fn append_frames(&mut self, frames: &mut Vec<virtio_video_format_frame>) {
+        self.frames.append(frames);
+        self.num_frames = (self.frames.len() as u32).into();
+    }
+
+    pub fn generate_mask(&mut self, num_formats: usize) {
+        self.mask = (u64::MAX >> (64 - num_formats)).into();
+    }
+}
+
+impl ToBytes for virtio_video_format_desc {
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut ret: Vec<u8> = Vec::new();
+
+        ret.extend_from_slice(self.mask.as_slice());
+        ret.extend_from_slice(self.format.as_slice());
+        ret.extend_from_slice(self.planes_layout.as_slice());
+        ret.extend_from_slice(self.plane_align.as_slice());
+        ret.extend_from_slice(self.num_frames.as_slice());
+        for frame in self.frames.iter() {
+            ret.append(&mut frame.to_bytes());
+        }
+        ret
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Default)]
+pub struct virtio_video_query_control_resp {
+    pub num: Le32,
+    pub padding: Le32,
+    pub value: Le32,
+}
+
+impl ToBytes for virtio_video_query_control_resp {
+    fn to_bytes(&self) -> Vec<u8> {
+        let mut ret: Vec<u8> = Vec::new();
+        ret.extend_from_slice(self.num.as_slice());
+        ret.extend_from_slice(self.padding.as_slice());
+        ret.extend_from_slice(self.value.as_slice());
```

Comment:
nit: Just chaining the iterators and collecting them into the vector would probably be faster since the correctly sized allocation could be done from the start.